packages/cli/src/Server.ts

```diff
@@ -1544,61 +1547,73 @@ class App {
 
-					const encryptionKey = await UserSettings.getEncryptionKey();
-					if (encryptionKey === undefined) {
-						throw new Error('No encryption key got found to encrypt the credentials!');
-					}
+				await validateEntity(newCredential);
 
-					if (incomingData.name === '') {
-						throw new Error('Credentials have to have a name set!');
-					}
+				// Add the added date for node access permissions
+				for (const nodeAccess of newCredential.nodesAccess) {
+					nodeAccess.date = this.getCurrentDate();
+				}
 
-					// Encrypt the data
-					const credentials = new Credentials(
-						{ id: null, name: incomingData.name },
-						incomingData.type,
-						incomingData.nodesAccess,
-					);
-					credentials.setData(incomingData.data, encryptionKey);
-					const newCredentialsData = credentials.getDataToSave() as ICredentialsDb;
+				const encryptionKey = await UserSettings.getEncryptionKey();
 
-					await this.externalHooks.run('credentials.create', [newCredentialsData]);
+				if (!encryptionKey) {
+					throw new Error('No encryption key was found to encrypt the credential!');
+				}
 
-					// Save the credentials in DB
-					const result = await Db.collections.Credentials!.save(newCredentialsData);
-					result.data = incomingData.data;
+				// Encrypt the data
+				const coreCredential = new Credentials(
+					{ id: null, name: newCredential.name },
+					newCredential.type,
+					newCredential.nodesAccess,
+				);
 
-					try {
-						await UserManagementHelpers.saveCredentialOwnership(
-							result as CredentialsEntity,
-							req.user as User,
-						);
-					} catch (error) {
-						LoggerProxy.error('Failed saving credential ownership.');
-					}
+				// @ts-ignore
+				coreCredential.setData(newCredential.data, encryptionKey);
 
-					// Convert to response format in which the id is a string
-					(result as unknown as ICredentialsResponse).id = result.id.toString();
-					return result as unknown as ICredentialsResponse;
-				},
-			),
+				const encryptedData = coreCredential.getDataToSave() as ICredentialsDb;
+
+				Object.assign(newCredential, encryptedData);
+
+				await this.externalHooks.run('credentials.create', [encryptedData]);
+
+				let savedCredential: undefined | CredentialsEntity;
+
+				await getConnection().transaction(async (transactionManager) => {
+					savedCredential = await transactionManager.save<CredentialsEntity>(newCredential);
+
+					savedCredential.data = newCredential.data;
+
+					const role = await Db.collections.Role!.findOneOrFail({
+						name: 'owner',
+						scope: 'credential',
+					});
+
+					const newSharedCredential = new SharedCredentials();
+
+					Object.assign(newSharedCredential, {
+						role,
+						user: req.user,
+						credentials: savedCredential,
+					});
+
+					await transactionManager.save<SharedCredentials>(newSharedCredential);
+				});
+
+				if (!savedCredential) {
+					throw new ResponseHelper.ResponseError('Failed to save credential', undefined, 500);
+				}
```

Comment:
I think if the transaction fails we anyway never reach this code 🤔

Reply:
Found a third way - returning a value inside the transaction makes it available in the outer scope. Otherwise TS would be unable to tell that `savedCredential` is not `undefined` after the transaction.